### PR TITLE
Multithreading todo parsing in TodoVisitor.

### PR DESCRIPTION
**Pull request by criske**
Just in case we'll need a multithreading aproach to parse todos.

### CODE REVIEW

**Comment by charlesmike** [comment]
@criske thank you for your Pull Request. I'll assign someone to review it soon.

File: src/main/java/com/selfxdsd/todocli/TodoParser.java [hunk 1]
```diff
@@ -77,13 +73,16 @@ public TodoParser() {
      * @param path Path to the file being parsed.
      * @return List of found TODOs.
      * @throws IOException If something goes wrong.
+     * @checkstyle ExecutableStatementCount (50 lines)
      */
     public List<Todo> parse(final String path) throws IOException {
         List<String> lines = Files.readAllLines(Paths.get(path));
         List<Todo> todos = new ArrayList<>();
+        final AtomicInteger currentIndex = new AtomicInteger();
```
Comment (criske) [inline src/main/java/com/selfxdsd/todocli/TodoParser.java]: @amihaiemil I've used a here an atomic int to be able to pass currentIndex to `parseMultiLineTodo` as a mutable ref. Before `currentIndex` was a class member and thus not thread safe.

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java [hunk 2]
```diff
@@ -102,25 +116,33 @@ public FileVisitResult visitFile(
...
+                        String suffix;
+                        if (i == todos.size() - 1) {
+                            suffix = "\n";
+                        } else {
+                            suffix = "";
+                        }
+                        log(todo.toString() + suffix);
+                    }
+
+                    this.serializer.addAll(todos);
```
Comment (criske) [inline src/main/java/com/selfxdsd/todocli/TodoVisitor.java:140]: @amihaiemil the serializer impl is already thread safe.

**Comment by charlesmike** [comment]
@amihaiemil I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

File: src/main/java/com/selfxdsd/todocli/TodoParser.java [hunk 3]
```diff
@@ -77,13 +73,16 @@ public TodoParser() {
...
 
-        for (currentIndex = 0; currentIndex < lines.size(); currentIndex++) {
-            String line = lines.get(currentIndex);
+        for (; currentIndex.get() < lines.size();
```
Comment (amihaiemil) [inline src/main/java/com/selfxdsd/todocli/TodoParser.java]: @criske This ``for`` looks really strange. Can we use a ``while`` instead?

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java [hunk 4]
```diff
@@ -89,8 +97,14 @@ public FileVisitResult postVisitDirectory(
             final Path dir, final IOException exc
     ) throws IOException {
         if (dir.equals(root)) {
-            // scanning root has finished.
-            this.serializer.serialize();
+            try {
+                // scanning root has finished.
+                service.shutdown();
+                service.awaitTermination(5, TimeUnit.MINUTES);
```
Comment (amihaiemil) [inline src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: @criske Why do we wait for 5 minutes here?

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java [hunk 5]
```diff
@@ -102,25 +116,33 @@ public FileVisitResult visitFile(
...
-
-            for (int i = 0; i < todos.size(); i++) {
-                Todo todo = todos.get(i);
-
-                String suffix;
-                if (i == todos.size() - 1) {
-                    suffix = "\n";
-                } else {
-                    suffix = "";
+            service.submit(() -> {
```
Comment (amihaiemil) [inline src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: @criske So, for each source code file, we start a separate thread?

**Review by amihaiemil — CHANGES_REQUESTED** [review]
@criske see questions :D

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java (hunk 4 above)
Comment (criske) [reply src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: That's the timeout period, basically it says the service should finish its running and queued jobs within 5 minutes. Most likely we'll never reach that timeout.

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java (hunk 5 above)
Comment (criske) [reply src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: Nope, as per service config, at most 3 threads, rest of the files will be queued.
`ExecutorService service = Executors.newFixedThreadPool(3)`

File: src/main/java/com/selfxdsd/todocli/TodoVisitor.java [hunk 6]
```diff
@@ -89,8 +105,19 @@ public FileVisitResult postVisitDirectory(
...
-            this.serializer.serialize();
+            try {
+                // scanning root has finished.
+                this.service.shutdown();
+                this.service.awaitTermination(5, TimeUnit.MINUTES);
+                if (this.hasErrors.get()) {
+                    throw new IOException("Could not serialize todos; something"
+                        + " went wrong while parsing a file for todos.");
+                } else {
+                    this.serializer.serialize();
```
Comment (criske) [inline src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: @amihaiemil now serialization is done only if there are no errors while parsing the src files

Comment (amihaiemil) [reply src/main/java/com/selfxdsd/todocli/TodoVisitor.java]: @criske no, please remove this and we'll do it in separate ticket if necessary. I want the parsing to take place even if there is an error in one of the todos. 

That's one of the things I don't like about the original [pdd](https://github.com/yegor256/pdd) -- it crashes and reports no tickets if it finds an error, even if only 1 todo is wrong.

**Comment by amihaiemil** [comment]
@rultor merge it

**Comment by amihaiemil** [comment]
@criske thanks! Paid with transaction id 96J5092377312730E

**Pull request merged**